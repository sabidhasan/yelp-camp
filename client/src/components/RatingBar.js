import React from 'react'
import PropTypes from 'prop-types'

class RatingBar extends React.Component {
  constructor(props) {
    super(props);
    this.handleKeyboard = this.handleKeyboard.bind(this);
    this.makeRatingArray = this.makeRatingArray.bind(this);
  }

  handleKeyboard(e) {
    if (!this.props.updateRating) return;
    if (e.key === 'ArrowRight' && this.props.rating < 5) {
      this.props.updateRating(this.props.rating + 1);
    } else if (e.key === 'ArrowLeft' && this.props.rating > 0) {
      this.props.updateRating(this.props.rating - 1);
    }
  }

  componentDidMount() {
    const elem = document.querySelector('.RatingBar--editable');
    if (!elem) return;
    elem.addEventListener('keydown', this.handleKeyboard);
  }

  componentWillUnmount() {
    const elem = document.querySelector('.RatingBar--editable');
    if (!elem) return;
    elem.removeEventListener('keydown', this.handleKeyboard);
  }

  makeRatingArray(initRating) {
    // for rating of campgrounds
    let rating = new Array(5);

    for (var i = 0; i < rating.length; i++) {
      if (initRating <= 0) {
        rating[i] = 0;
      } else if (initRating < 100) {
        rating[i] = Math.round(initRating / 10) * 10;
        initRating = 0;
      } else {
        rating[i] = 100;
        initRating = Math.round(initRating - 100);
      }
    }
    return rating
  }

  render() {
    const initRating = (Math.round(this.props.rating * 10) * 10) || 0;
    let rating = this.makeRatingArray(initRating);
    rating = rating.map((val, idx) => (
        <span
          aria-label={`${idx+1} Star`}
          aria-checked={val !== 0 ? 'true' : 'false'}
          role='radio'
          key={idx}
          onClick={() => {if (this.props.updateRating) this.props.updateRating(idx + 1)}}
          className={`RatingBar__star star${val}` +
              `${this.props.small ? ' RatingBar__star--small' : ''}`
          }
        >
        </span>
    ));
    return (
      <div
        className={`RatingBar ${this.props.updateRating ? ' RatingBar--editable' : ''}`}
        aria-label={`Rating: ${this.props.rating} stars out of 5. ${this.props.updateRating ? 'Rating selectable with arrow keys' : ''} `}
        role='radiogroup'
        tabIndex={this.props.updateRating ? '0' : null}
      >
        {rating}
      </div>
    )
  }
}

RatingBar.propTypes = {
  updateRating: PropTypes.func,
  rating: PropTypes.number,
  small: PropTypes.bool
}

export default RatingBar
