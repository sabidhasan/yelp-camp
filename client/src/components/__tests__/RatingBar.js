import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import RatingBar from '../RatingBar'

configure({ adapter: new Adapter() });

describe('<RatingBar />', () => {
  let component;
  let props;
  let updateRating;

  beforeEach(() => {
    updateRating = jest.fn();
    props = {
      updateRating,
      small: true,
      rating: 3
    }
    component = shallow(<RatingBar { ...props } />)
  })

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  })

  it('Right arrow key with rating < 5 calls updateRating function with new rating', () => {
    component.instance().handleKeyboard({key: 'ArrowRight'})
    expect(updateRating).toBeCalledWith(4)
  });

  it('Right arrow key with rating = 5 does not call updateRating function', () => {
    // Update the rating
    component.setProps({rating: 5})
    component.instance().handleKeyboard({key: 'ArrowRight'})
    expect(updateRating).toHaveBeenCalledTimes(0);
  });

  it('Left arrow key with rating > 0 calls updateRating function with new rating', () => {
    component.instance().handleKeyboard({key: 'ArrowLeft'})
    expect(updateRating).toBeCalledWith(2)
  });

  it('Left arrow key with rating of <= 0 does not call updateRating function', () => {
    // Update the rating
    component.setProps({rating: 0})
    component.instance().handleKeyboard({key: 'ArrowLeft'})
    expect(updateRating).toHaveBeenCalledTimes(0);
  });

  it('Ratingbar--editable exists', () => {
    expect(component.find('.RatingBar--editable')).toHaveLength(1)
  })

  it('Ratingbar--editable does not exist when updateRating is not supplied', () => {
    component.setProps({updateRating: null})
    expect(component.find('.RatingBar--editable')).toHaveLength(0)
  })

  it('RatingBar__star--small is shown when small prop passed', () => {
    expect(component.find('.RatingBar__star--small')).toHaveLength(5)
  });

  it('RatingBar__star--small is hidden when small prop not passed', () => {
    component.setProps({small: false})
    expect(component.find('.RatingBar__star--small')).toHaveLength(0)
  });

  it('Clicking RatingBar__star calls the updateRating function', () => {
    component.find('.RatingBar__star').at(3).simulate('click');
    expect(updateRating).toBeCalledWith(4)
  });

  it('Rating of 2 produces two .star100 elements, and 3 .star0 elements', () => {
    component.setProps({rating: 2})
    expect(component.find('.star100')).toHaveLength(2);
    expect(component.find('.star0')).toHaveLength(3);
  })

  it('Rating of 0 produces five .star0 elements', () => {
    component.setProps({rating: 0})
    expect(component.find('.star0')).toHaveLength(5);
  })

  it('Decimal rating produces proper elements (no rouding errors)', () => {
    component.setProps({rating: 3.4});
    expect(component.find('.star100')).toHaveLength(3);
    expect(component.find('.star40')).toHaveLength(1);
  })

  it('Greater than 5 rating rounds off to 5 stars', () => {
    component.setProps({rating: 9.4});
    expect(component.find('.star100')).toHaveLength(5);
  })

  it('Negative rating shows as 0 stars', () => {
    component.setProps({rating: -9.4});
    expect(component.find('.star0')).toHaveLength(5);
  })

  it('Illogically precise rating produces proper elements', () => {
    component.setProps({rating: 4.4238722});
    expect(component.find('.star100')).toHaveLength(4);
    expect(component.find('.star40')).toHaveLength(1);
  })

  it('RatingBar component snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  })
});
