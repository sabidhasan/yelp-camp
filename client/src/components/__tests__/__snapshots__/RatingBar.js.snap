// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RatingBar /> RatingBar component snapshot matches what is expected 1`] = `
<div
  aria-label="Rating: 3 stars out of 5. Rating selectable with arrow keys "
  className="RatingBar  RatingBar--editable"
  role="radiogroup"
  tabIndex="0"
>
  <span
    aria-checked="true"
    aria-label="1 Star"
    className="RatingBar__star star100 RatingBar__star--small"
    key="0"
    onClick={[Function]}
    role="radio"
  />
  <span
    aria-checked="true"
    aria-label="2 Star"
    className="RatingBar__star star100 RatingBar__star--small"
    key="1"
    onClick={[Function]}
    role="radio"
  />
  <span
    aria-checked="true"
    aria-label="3 Star"
    className="RatingBar__star star100 RatingBar__star--small"
    key="2"
    onClick={[Function]}
    role="radio"
  />
  <span
    aria-checked="false"
    aria-label="4 Star"
    className="RatingBar__star star0 RatingBar__star--small"
    key="3"
    onClick={[Function]}
    role="radio"
  />
  <span
    aria-checked="false"
    aria-label="5 Star"
    className="RatingBar__star star0 RatingBar__star--small"
    key="4"
    onClick={[Function]}
    role="radio"
  />
</div>
`;
